src/eventsub/event.rs

```diff
@@ -52,111 +52,134 @@ macro_rules! is_thing {
-    /// `channel.subscription.gift`: a user sends a resubscription chat message in a specific channel
-    #[serde(rename = "channel.subscription.message")]
-    ChannelSubscriptionMessage,
-    /// `channel.goal.begin`: a goal begins on the specified channel.
-    #[serde(rename = "channel.goal.begin")]
-    ChannelGoalBegin,
-    /// `channel.goal.progress`: a goal makes progress on the specified channel.
-    #[serde(rename = "channel.goal.progress")]
-    ChannelGoalProgress,
-    /// `channel.goal.end`: a goal ends on the specified channel.
-    #[serde(rename = "channel.goal.end")]
-    ChannelGoalEnd,
-    /// `channel.hype_train.begin`: a hype train begins on the specified channel.
-    #[serde(rename = "channel.hype_train.begin")]
-    ChannelHypeTrainBegin,
-    /// `channel.hype_train.progress`: a hype train makes progress on the specified channel.
-    #[serde(rename = "channel.hype_train.progress")]
-    ChannelHypeTrainProgress,
-    /// `channel.hype_train.end`: a hype train ends on the specified channel.
-    #[serde(rename = "channel.hype_train.end")]
-    ChannelHypeTrainEnd,
-    /// `stream.online`: the specified broadcaster starts a stream.
-    #[serde(rename = "stream.online")]
-    StreamOnline,
-    /// `stream.online`: the specified broadcaster stops a stream.
-    #[serde(rename = "stream.offline")]
-    StreamOffline,
-    /// `user.update`: user updates their account.
-    #[serde(rename = "user.update")]
-    UserUpdate,
-    /// `user.authorization.revoke`: a user has revoked authorization for your client id. Use this webhook to meet government requirements for handling user data, such as GDPR, LGPD, or CCPA.
-    #[serde(rename = "user.authorization.revoke")]
-    UserAuthorizationRevoke,
-    /// `user.authorization.revoke`: a user’s authorization has been granted to your client id.
-    #[serde(rename = "user.authorization.grant")]
-    UserAuthorizationGrant,
+macro_rules! make_event_type {
+    ($enum_docs:literal: pub enum $enum_name:ident {
+        $(
+            $event_docs:literal:
+            $variant_name:ident => $event_name:literal,
+        )*
+    },
+        to_str: $to_str_docs:literal
+    ) => {
+        #[doc = $enum_docs]
+        #[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+        #[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
+        #[non_exhaustive]
+        pub enum $enum_name {
+            $(
+                #[doc = concat!("`", $event_name, "`: ", $event_docs)]
+                #[serde(rename = $event_name)]
+                $variant_name,
+            )*
+        }
+
+        impl $enum_name {
+            #[doc = $to_str_docs]
+            pub const fn to_str(&self) -> &'static str {
+                use $enum_name::*;
+                match self {
+                    $($variant_name => $event_name,)*
+                }
+            }
+        }
+
+        impl std::str::FromStr for $enum_name {
+            type Err = ();
```

Comment:
this could be `std::convert::Infallible`

Reply:
Oh wait it can't, thought I had a custom eventtype here, we do not.

Make a simple error

```rust
#[derive(thiserror::Error, Debug, Clone)]
#[error("Unknown event type: {0})]
pub struct EventTypeError(String);
```

and use that

Reply:
I don't think we should allocate for the error. The standard library for example uses a generic error message when parsing an ip address: https://doc.rust-lang.org/src/std/net/parser.rs.html#365-388, here something like this would be similar:

```rust
#[derive(thiserror::Error, Debug, Clone)]
#[error("Unknown event type")]
pub struct EventTypeParseError;
```

It's probably done like this because the underlying string isn't moved thus users still have the source string.

Reply:
sounds good!